spec/lib/nerve/reporter_zookeeper_spec.rb

```diff
@@ -29,5 +31,103 @@
     reporter.report_up
     reporter.stop
   end
+
+  context "when reporter has been up and there is connection issue" do
+    before(:each) do
+      allow(zk).to receive(:close!)
+      allow(zk).to receive(:connected?).and_return(true)
+      expect(zk).to receive(:exists?) { "zk_path" }.and_return(false)
+      expect(zk).to receive(:mkdir_p) { "zk_path" }
+      expect(zk).to receive(:create) { "full_path" }
+      allow(ZK).to receive(:new).and_return(zk)
+
+      @reporter = Nerve::Reporter::Zookeeper.new(subject)
+      @reporter.start
+      @reporter.report_up
+    end
+
+    after(:each) do
+      # reset the class variable to avoid mock object zk leak
+      Nerve::Reporter::Zookeeper.class_variable_set(:@@zk_pool, {})
+      Nerve::Reporter::Zookeeper.class_variable_set(:@@zk_pool_count, {})
+    end
+
+    it 'returns false on ping? when zk is disconnected' do
+      # this condition is triggered if connection has been lost for a while (a few sec)
+      expect(zk).to receive(:connected?).and_return(false)
+      expect(zk).not_to receive(:exists?)
+      expect(@reporter.ping?).to be false
+    end
+
+    it 'returns false on ping? when zk.exists? check failed due to connection loss' do
```

Comment:
great example where you missuse the `it` block to set up a context. this would be better as:

```ruby
context 'when zk.exists? check fails to due to connection loss' do
  before { expect(zk).to receive(:exists?).and_raise(ZK::Exceptions::OperationTimeOut) }

  it 'returns false on ping' do
    expect(@reporter.ping?).to be_falsey
  end
end
```